src/rgw/driver/rados/rgw_datalog.cc

```diff
@@ -1080,14 +1499,238 @@ std::string RGWDataChangesLog::max_marker() const {
+	break;
+      } else {
+	throw;
+      }
+    }
+  } while (!cursor.empty());
+  co_return;
+}
+
+asio::awaitable<bool>
+RGWDataChangesLog::synthesize_entries(
+  const DoutPrefixProvider* dpp,
+  int index,
+  const std::unordered_map<std::string, uint64_t>& semcount)
+{
+  const auto timestamp = real_clock::now();
+  auto be = bes->head();
+  auto push_failed = false;
+
+  RGWDataChangesBE::entries batch;
+  for (const auto& [key, sem] : semcount) {
+    try {
+      BucketGen bg{key};
+      rgw_data_change change;
+      buffer::list bl;
+      change.entity_type = ENTITY_TYPE_BUCKET;
+      change.key = bg.shard.get_key();
+      change.timestamp = timestamp;
+      change.gen = bg.gen;
+      encode(change, bl);
+      be->prepare(timestamp, change.key, std::move(bl), batch);
+    } catch (const sys::error_code& e) {
+      push_failed = true;
+      ldpp_dout(dpp, -1) << "RGWDataChangesLog::synthesize_entries(): Unable to "
+			 << "parse Bucketgen key: " << key << "Got exception: "
+			 << e.what() << dendl;
+    }
+  }
+  try {
+    co_await be->push(dpp, index, std::move(batch));
+  } catch (const std::exception& e) {
+    push_failed = true;
+    ldpp_dout(dpp, 5) << "RGWDataChangesLog::synthesize_entries(): Backend push "
+		      << "failed with exception: " << e.what() << dendl;
+  }
+  co_return !push_failed;
+}
+
+asio::awaitable<bool>
+RGWDataChangesLog::gather_working_sets(
+  const DoutPrefixProvider* dpp,
+  int index,
+  std::unordered_map<std::string, uint64_t>& semcount)
+{
+  buffer::list bl;
+  recovery_check rc = index;
+  encode(rc, bl);
+  auto [reply_map, missed_set] = co_await rados->notify(
+      get_sem_set_oid(0), loc, bl, 60s, asio::use_awaitable);
+  // If we didn't get an answer from someone, don't decrement anything.
+  if (!missed_set.empty()) {
+    ldpp_dout(dpp, 5) << "RGWDataChangesLog::gather_working_sets(): Missed responses: "
+		      << missed_set << dendl;
+    co_return false;
+  }
+  for (const auto& [source, reply] : reply_map) {
+    std::unordered_set<std::string> keys;
+    try {
+      decode(keys, reply);
+    } catch (const std::exception& e) {
+      ldpp_dout(dpp, -1)
+	<< "RGWDataChangesLog::gather_working_sets(): Failed decoding reply from: "
+	<< source << dendl;
+      co_return false;
+    }
+    for (const auto& key : keys) {
+      auto iter = semcount.find(key);
+      if (iter == semcount.end()) {
+	continue;
+      }
+      if (iter->second == 1) {
+	semcount.erase(iter);
+      } else {
+	--(iter->second);
+      }
+    }
+  }
+  co_return true;
+}
+
+asio::awaitable<void>
+RGWDataChangesLog::decrement_sems(
+  int index,
+  std::unordered_map<std::string, uint64_t>&& semcount)
+{
+  namespace sem_set = neorados::cls::sem_set;
+  while (!semcount.empty()) {
+    std::unordered_set<std::string> batch;
+    for (auto j = 0u; j < sem_max_keys && !semcount.empty(); ++j) {
+      auto iter = std::begin(semcount);
+      batch.insert(iter->first);
+      semcount.erase(iter);
```

Comment:
might as well std::move this string since you're erasing it right away

Reply:
Done